056-kafka-roller-kraft.md

```diff
@@ -0,0 +1,129 @@
+is implemented in order to check the quorum state when rolling controllers at least one broker needs to be running.
+
+### Triggers
+The following will trigger an attempt to roll a KRaft controller or combined node:
+- Pod is annotated for manual rolling update
+- Pod is unready
+- Controller's configuration has changed and cannot be updated dynamically.
+
+The triggers for broker remain the same as Zookeeper mode.
+
+**NOTE** There is a related proposal describing how to diff controller configuration currently being discussed: [PR #82](https://github.com/strimzi/proposals/pull/82)
+
+### Rollability checks
+
+The checks made by the KafkaRoller in different modes is described below.
+The checks include a new check for controllers to verify that rolling the node does not affect the quorum health.
+The proposed check is:
+- Fetch `lastCaughtUpTimestamp` for every node
+- A node is marked as caught up if `leaderLastCaughtUpTimestamp - lastCaughtUpTimestamp < controllerQuorumFetchTimeoutMs`
+- Count each node that is caught up, plus the leader (`numOfCaughtUpVoters`)
+- Can roll if: `numOfCaughtUpVoters > (numOfVoters)/2`
+
+#### Separate brokers and controllers
+For controller-only:
+- Does not force roll a controller performing log recovery
+- Does not roll a ready controller if doing so would take the number of caught-up voters (inc leader) to less than half of the quorum
+- When rolling an unready controller, does not check quorum state before rolling
+
+For broker-only:
+- Attempts to connect an admin client to the broker and if it can't connect rolls the broker
+- Does not force roll a broker performing log recovery
+- Does not roll a broker if doing so would take the in-sync replicas count below `min.insync.replicas`
+- **NEW** If a broker is not ready, check the quorum is formed before force rolling the broker
+
+#### Combined mode
+- Does not force roll a combined node performing log recovery
+- Does not roll a ready combined node if doing so would take the number of caught-up voters (inc leader) to less than half of the quorum
+- Does not roll a ready combined node if doing so would take the in-sync replicas count below `min.insync.replicas`
+- When rolling an unready combined node, does not check quorum state, or in-sync replicas count, or try to connect an admin client to the broker before rolling
+
+### Configuration changes
+For all kinds of nodes:
+- Compares the current configuration with expected configuration to see if a node needs rolling
+- Performs dynamic configuration updates if possible, otherwise rolls node on configuration change
+
+**NOTE:** At the time of writing this proposal dynamic configuration updates are not possible for controller nodes.
+However [KIP 919](https://cwiki.apache.org/confluence/display/KAFKA/KIP-919%3A+Allow+AdminClient+to+Talk+Directly+with+the+KRaft+Controller+Quorum) looks to address this
```

Comment:
So, what does that mean for the proposal? Does it expect to wait for it to be implemented? If not what workarounds / limitations will it take?

Reply:
I tried making this little more clear. Some part would be resolved by the workaround and some part would not be done until this KIP is implemented.